cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/RemoteAppStateStoreTest.java

```diff
@@ -0,0 +1,224 @@
+import io.cdap.cdap.internal.app.store.state.AppStateKey;
+import io.cdap.cdap.proto.NamespaceMeta;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.security.auth.context.AuthenticationTestContext;
+import io.cdap.http.NettyHttpService;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.InMemoryDiscoveryService;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * Tests for {@link RemoteAppStateStore}
+ */
+public class RemoteAppStateStoreTest {
+
+  private static final String NAMESPACE = "ns1";
+
+  private static NettyHttpService httpService;
+  private static ApplicationLifecycleService applicationLifecycleService;
+  private static CConfiguration cConf;
+  private static RemoteClientFactory remoteClientFactory;
+  private static Cancellable cancellable;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    cConf = CConfiguration.create();
+    cConf.set("app.state.retry.policy.base.delay.ms", "10");
+    cConf.set("app.state.retry.policy.max.delay.ms", "2000");
+    cConf.set("app.state.retry.policy.max.retries", "2147483647");
+    cConf.set("app.state.retry.policy.max.time.secs", "60");
+    cConf.set("app.state.retry.policy.type", "exponential.backoff");
+    InMemoryDiscoveryService discoveryService = new InMemoryDiscoveryService();
+    NamespaceAdmin namespaceAdmin = new InMemoryNamespaceAdmin();
+    applicationLifecycleService = Mockito.mock(ApplicationLifecycleService.class);
+    NamespaceMeta testNameSpace = new NamespaceMeta.Builder()
+      .setName(NAMESPACE)
+      .setDescription("This is the default namespace, which is automatically created, and is always available.")
+      .build();
+    namespaceAdmin.create(testNameSpace);
+    remoteClientFactory = new RemoteClientFactory(discoveryService,
+                                                  new DefaultInternalAuthenticator(new AuthenticationTestContext()));
+    httpService = new CommonNettyHttpServiceBuilder(cConf, "appfabric", new NoOpMetricsCollectionService())
+      .setHttpHandlers(new AppStateHandler(applicationLifecycleService, namespaceAdmin)).build();
+    httpService.start();
+    cancellable = discoveryService
+      .register(URIScheme.createDiscoverable(Constants.Service.APP_FABRIC_HTTP, httpService));
+  }
+
+  @AfterClass
+  public static void cleanUp() throws Exception {
+    cancellable.cancel();
+    httpService.stop();
+  }
+
+  @Test
+  public void testSaveSuccess() {
+    RemoteAppStateStore remoteAppStateStore = new RemoteAppStateStore(cConf, remoteClientFactory, NAMESPACE, "app1");
+    byte[] value = "testvalue".getBytes();
+    try {
+      remoteAppStateStore.saveState("key1", value);
+    } catch (IOException e) {
+      //expected
```

Comment:
it's not expected right? Seems like it can just propagate the exception up instead of catching and failing

Reply:
similar comment for the other tests

Reply:
Done